runtime-sdk/src/modules/rofl/mod.rs

```diff
@@ -0,0 +1,535 @@
+    Runtime,
+};
+
+pub mod app;
+pub mod app_id;
+mod config;
+mod error;
+mod event;
+pub mod policy;
+pub mod state;
+#[cfg(test)]
+mod test;
+pub mod types;
+
+/// Unique module name.
+const MODULE_NAME: &str = "rofl";
+
+pub use config::Config;
+pub use error::Error;
+pub use event::Event;
+
+/// Parameters for the module.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct Parameters {}
+
+/// Errors emitted during parameter validation.
+#[derive(thiserror::Error, Debug)]
+pub enum ParameterValidationError {}
+
+impl module::Parameters for Parameters {
+    type Error = ParameterValidationError;
+
+    fn validate_basic(&self) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
+
+/// Genesis state for the module.
+#[derive(Clone, Debug, Default, cbor::Encode, cbor::Decode)]
+pub struct Genesis {
+    pub parameters: Parameters,
+
+    /// Application configurations.
+    pub apps: Vec<types::AppConfig>,
+}
+
+/// Interface that can be called from other modules.
+pub trait API {
+    /// Verify whether the origin transaction is signed by an authorized ROFL instance for the given
+    /// application.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if called outside a transaction environment.
+    fn is_authorized_origin(app: app_id::AppId) -> Result<bool, Error>;
+
+    /// Get an application's configuration.
+    fn get_app(id: app_id::AppId) -> Result<types::AppConfig, Error>;
+
+    /// Get all registered instances for an application.
+    fn get_instances(id: app_id::AppId) -> Result<Vec<types::Registration>, Error>;
+}
+
+/// Module's address that has the application stake pool.
+///
+/// oasis1qza6sddnalgzexk3ct30gqfvntgth5m4hsyywmff
+pub static ADDRESS_APP_STAKE_POOL: Lazy<Address> =
+    Lazy::new(|| Address::from_module(MODULE_NAME, "app-stake-pool"));
+
+pub struct Module<Cfg: Config> {
+    _cfg: std::marker::PhantomData<Cfg>,
+}
+
+impl<Cfg: Config> API for Module<Cfg> {
+    fn is_authorized_origin(app: app_id::AppId) -> Result<bool, Error> {
+        let caller_pk = CurrentState::with_env_origin(|env| env.tx_caller_public_key())
+            .ok_or(Error::InvalidArgument)?;
+
+        // Resolve RAK as the call may be made by an extra key.
+        let rak = match state::get_endorser(&caller_pk) {
+            // It may point to a RAK.
+            Some(state::KeyEndorsementInfo { rak: Some(rak), .. }) => rak,
+            // Or it points to itself.
+            Some(_) => caller_pk.try_into().map_err(|_| Error::InvalidArgument)?,
+            // Or is unknown.
+            None => return Ok(false),
+        };
+
+        // We need to also fetch the registration to ensure it is registered for the right app.
+        let registration = match state::get_registration(app, &rak) {
+            Some(registration) => registration,
+            None => return Ok(false),
+        };
+
+        // Ensure enclave is registered for the correct app.
```

Comment:
These checks can now be removed (I think there are 2 of them).

Reply:
Where is the other one?

Reply:
Ups, only one.

Reply:
Fixed.